plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunctionProcessor.java

```diff
@@ -0,0 +1,212 @@
+            return TableFunctionProcessorState.Processed.usedInputAndProduced(new Page(page.getPositionCount(), resultBlock));
+        }
+        return FINISHED;
+    }
+
+    private TableFunctionProcessorState processDataFile()
+    {
+        Page page = deltaLakePageSource.getNextPage();
+        if (page != null) {
+            int filePageColumns = page.getChannelCount();
+            Block[] blocks = new Block[filePageColumns + NUMBER_OF_ADDITIONAL_COLUMNS_FOR_DATA_FILE];
+            for (int i = 0; i < filePageColumns; i++) {
+                blocks[i] = page.getBlock(i);
+            }
+            blocks[filePageColumns] = RunLengthEncodedBlock.create(
+                    nativeValueToBlock(VARCHAR, utf8Slice("insert")), page.getPositionCount());
+            blocks[filePageColumns + 1] = RunLengthEncodedBlock.create(
+                    currentVersionAsBlock, page.getPositionCount());
+            blocks[filePageColumns + 2] = RunLengthEncodedBlock.create(
+                    nativeValueToBlock(TIMESTAMP_TZ_MILLIS, packDateTimeWithZone(currentVersionCommitTimestamp, UTC_KEY)), page.getPositionCount());
+            return TableFunctionProcessorState.Processed.usedInputAndProduced(new Page(page.getPositionCount(), blocks));
+        }
+        return FINISHED;
+    }
+
+    private DeltaLakePageSource createDeltaLakePageSource(TableChangesSplit split)
+    {
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+        TrinoInputFile inputFile = fileSystem.newInputFile(split.path());
+        Map<String, Optional<String>> partitionKeys = split.partitionKeys();
+
+        try {
+            ReaderPageSource pageSource = ParquetPageSourceFactory.createPageSource(
+                    inputFile,
+                    0,
+                    split.fileSize(),
+                    splitColumns.stream().filter(column -> column.getColumnType() == REGULAR).map(DeltaLakeColumnHandle::toHiveColumnHandle).collect(toImmutableList()),
+                    TupleDomain.all(), // TODO add predicate pushdown https://github.com/trinodb/trino/issues/16990
+                    true,
+                    parquetDateTimeZone,
+                    fileFormatDataSourceStats,
+                    parquetReaderOptions,
+                    Optional.empty(),
+                    domainCompactionThreshold);
+
+            verify(pageSource.getReaderColumns().isEmpty(), "Unexpected reader columns: %s", pageSource.getReaderColumns().orElse(null));
+
+            return new DeltaLakePageSource(
+                    splitColumns,
+                    ImmutableSet.of(),
+                    partitionKeys,
+                    Optional.empty(),
+                    pageSource.get(),
+                    split.path(),
+                    split.fileSize(),
+                    Optional.empty());
+        }
+        catch (TrinoException e) {
+            throw new TrinoException(DELTA_LAKE_BAD_DATA, "Failed to access table data");
```

Comment:
I don't wrap anything here, I just throw this one instead of  `new TrinoException(HIVE_CANNOT_OPEN_SPLIT, message, e);`

Reply:
well, you _should_ wrap, otherwise cause of exception (actual problem) is lost (eg how can i know this is permission denied issue?)

i am fine throwing current `TrinoException`, without translating.